src/theory/quantifiers/ematching/pattern_term_selector.h

```diff
@@ -0,0 +1,190 @@
+   * is a usable trigger for q if one exists, otherwise returns Node::null().
+   */
+  static Node getIsUsableEq(Node q, Node eq);
+  /** returns whether n1 == n2 is a usable (relational) trigger for q. */
+  static bool isUsableEqTerms(Node q, Node n1, Node n2);
+  /** Helper for collect, with a fixed strategy for selection and filtering */
+  void collectInternal(Node n,
+                       std::vector<Node>& patTerms,
+                       std::map<Node, TriggerTermInfo>& tinfo,
+                       options::TriggerSelMode tstrt,
+                       bool filterInst);
+  /** recursive helper function for collectPatTerms
+   *
+   * This collects the usable trigger terms in the subterm n of the body of
+   * quantified formula of this class.
+   *   visited : cache of the trigger terms collected for each visited node,
+   *   tinfo : cache of trigger term info for each visited node,
+   *   tstrat : the selection strategy (see options/quantifiers_mode.h)
+   *   pol/hasPol : the polarity of node n in q
+   *                (see QuantPhaseReq theory/quantifiers/quant_util.h)
+   *   epol/hasEPol : the entailed polarity of node n in q
+   *                  (see QuantPhaseReq theory/quantifiers/quant_util.h)
+   *   knowIsUsable : whether we know that n is a usable trigger.
+   *
+   * We add the triggers we collected recursively in n into added.
+   */
+  void collectTermsInternal(Node n,
+                            std::map<Node, std::vector<Node> >& visited,
+                            std::map<Node, TriggerTermInfo>& tinfo,
+                            options::TriggerSelMode tstrt,
+                            std::vector<Node>& added,
+                            bool pol,
+                            bool hasPol,
+                            bool epol,
+                            bool hasEPol,
+                            bool knowIsUsable = false);
+
+  /** filter all nodes that have instances
+   *
+   * This is used during collectModelInfo to filter certain trigger terms,
+   * stored in nodes. This updates nodes so that no pairs of distinct nodes
+   * (i,j) is such that i is a trigger instance of j or vice versa (see below).
+   */
+  static void filterInstances(std::vector<Node>& nodes);
+
+  /** is instance of
+   *
+   * We say a term t is an trigger instance of term s if
+   * (1) t = s * { x1 -> t1 ... xn -> tn }
+   * (2) { x1, ..., xn } are a subset of FV( t ).
+   * For example, f( g( h( x, x ) ) ) and f( g( x ) ) are instances of f( x ),
+   * but f( g( y ) ) and g( x ) are not instances of f( x ).
+   *
+   * When this method returns -1, n1 is an instance of n2,
+   * When this method returns 1, n1 is an instance of n2.
+   *
+   * The motivation for this method is to discard triggers s that are less
+   * restrictive (criteria (1)) and serve to bind the same variables (criteria
+   * (2)) as another trigger t. This often helps avoiding matching loops.
+   */
+  static int isInstanceOf(Node n1,
```

Comment:
For this guy's implementation can't you use the stand alone unification methods added recently?

Reply:
No, since the terms are in "instantiation constant" form (they don't have free variables)